### ISSUE TITLE
Refactoring of _add_paths function used to create graph

### ISSUE THREAD

**Issue by zacharyDez** [gboeing/osmnx#538]
**Problem being addressed**

The function _add_path was doing multiple tasks making it hard to grasp rapidly. It first evaluated if the path was a one-way given multiple conditions. Then it updated data by reversing it if the path was a one-way and if value of the key in osm path data was either -1 or T. The evaluation of the data being reversed was also within a single condition to avoid hitting a key error raised by the fact that the path data dictionary does not always contain the one-way key for the other conditions. Finally, the function _add_path, used to add a single path, was called after checking every condition if the path was a one-way.

**Proposed solution**

Extracting function _is_path_one_way to evaluate if the path is a one-way, returns a bool. 

Extracting function _is_path_data_reversed to evaluate if the path data is reversed using a try catch to simplify logic, returns a bool.

Reversing path data if both conditions above are true.

Make a single function call in the loop of _add_paths to _add_path passing as the one-way parameter the evaluated condition of _is_path_one_way.

Body of _add_paths after refactor
![](https://user-images.githubusercontent.com/36894613/88177339-b22e6600-cbf6-11ea-8efe-053f24586540.PNG)

**Comment by gboeing** [comment]
@zacharyDez thanks. Can you explain in a bit more detail the motivation for this proposal. Is there a specific bug you identified that this resolves? Does this improve the runtime or memory footprint? You extracted and created some new functions: are these used elsewhere in the codebase?

**Comment by zacharyDez** [comment]
@gboeing. The main improvement functionality wise is the reduced memory footprint. The graph was being copied with the add_paths function. The functions extracted are not used elsewhere in the codebase, only modified a single file.  Otherwise, the improvement is more about the readability of the code. As mentioned, the sequence of the add_paths function was hard to follow as mentioned above.

**Comment by gboeing** [comment]
Thank you! Memory footprint improvements are much appreciated. I'll have time again after Aug 1 to review PRs in detail. In the meantime could you share the memory use difference before/after the proposed PR? Thanks!
